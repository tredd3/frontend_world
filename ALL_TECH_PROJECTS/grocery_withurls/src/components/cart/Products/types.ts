export type ProductRenderVariations =
  'save-for-later' | 'cart' | 'cart-update' | 'review-cart';
