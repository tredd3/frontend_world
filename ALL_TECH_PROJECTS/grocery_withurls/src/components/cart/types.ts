import { CartPivot, ShipmentProductPivot } from '../../types';

export type OnCartUpdated = (cart?: CartPivot) => void;

export type ActionHandlers = {
  onSaveForLater: (product: ShipmentProductPivot) => Promise<void>;
  onMoveToCart: (product: ShipmentProductPivot) => Promise<void>;
  onDeleteSavedForLater: (product: ShipmentProductPivot) => Promise<void>;
  onDelete: (product: ShipmentProductPivot) => Promise<void>;
  onChangeQuantity: (product: ShipmentProductPivot, quantity: number) => Promise<void>;
}
