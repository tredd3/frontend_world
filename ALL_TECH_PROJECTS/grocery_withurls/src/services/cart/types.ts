import { actions } from './shared';

export type NonCheckoutParams = { action: typeof actions.preCheckout };
export type CheckoutParams = {
  action: typeof actions.postCheckout;
  // customerAddressId is actually a string in the request. Need to check if number works.
  // If not, we'll have to cast before sending it.
  customerAddressId: number;
  deliveryDate: string;
  timeslot: '';
  isAutoSelected: 0;
};

export type MaybeCheckout = CheckoutParams | NonCheckoutParams;

export type AddCartItemRequestBase = {
  userId: number;
  storeId?: number;
  cartId: number;
  product: { skuId: number; qty: number; productType: number }[];
};

export type AddCartItemRequest = AddCartItemRequestBase & NonCheckoutParams;

export type ChangeQuantityRequest = AddCartItemRequestBase & MaybeCheckout;

export type RemoveCartItemRequest = {
  userId: number;
  storeId?: number;
  cartId: number;
  skuId: number;
  qty: 0;
  type: number;
} & MaybeCheckout;

export type MoveCartItemRequest = {
  userId: number;
  cartId: number;
  skuId: number;
  qty: number;
  productType: number;
  moveType: 'SaveForLater' | 'MoveToCart';
};

export type DeleteSavedForLaterItemRequest = {
  skuId: number;
  qty: number;
  productType: number;
  productCategoryType: 'SaveForLater';
};

export type PaymentRequestRequest = {
  Amount: number;
  CartId: number;
  MobileNo: number;
  StoreId: number;
  UserId: number;
};

export type PaymentRequestResponse = {
  SUPERMID: string;
  CLIENTID: string;
  MERCHANTNAME: string;
  SPLITTXNFLAG: string;
  CHANNEL: string;
  CURRENCY: string;
  TXNTYPE: string;
  Time: string;
  Checksum: string;
  Amount: string;
  RefNo: string;
  SplitDetails: string;
  ReturnUrl: string;
  PaymentUrl: string;
  Version: string;
  IncludePayModes: string;
  ExcludePayModes: string;
  ProductDescription: string;
  UDF1: string;
  UDF2: string;
  UDF3: string;
  UDF4: string;
  UDF5: string;
};

export type PGPostPayload = {
  merchantid: string;
  clientid: string;
  merchantname: string;
  splittxnflag: string;
  channel: string;
  'transaction.currency': string;
  'transaction.txntype': string;
  'transaction.timestamp': string;
  checksum: string;
  'transaction.amount': string;
  'transaction.extref': string;
  'transaction.splitdetails': string;
  returl: string;
  version: string;
  'preferences.includepaymodes': string;
  'preferences.excludepaymodes': string;
  productdescription: string;
  udf1: string;
  udf2: string;
  udf3: string;
  udf4: string;
  udf5: string;
  token: string;
  'subscriber.customername': string;
  'subscriber.mobilenumber': string;
  'subscriber.customerid': string;
}

export type DeliveryModesAPIRequestType = {
  StoreId: number;
  OrderValue: number;
  Pincode: number;
};

export type CheckoutRequest = {
  storeId: number;
  cartId: number;
  userId: number;
  // Not sure if the user's mobile number is needed
  // userMobile: number;
} & CheckoutParams;

export type PersistedPaymentDetails = {
  refNo: string;
  userId: number;
  storeId: number;
  cartId: number;
};

export type PaymentResponseRequest = {
  RefNo: string;
  UserId: number;
  StoreId: number;
};

export type PlaceCODOrderRequest = {
  CartId: number;
  StoreId: number;
  PaymentMethod: 1;
  MID: 0;
}
