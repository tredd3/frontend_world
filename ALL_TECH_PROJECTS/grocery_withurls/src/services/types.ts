/* eslint-disable max-classes-per-file */
import { ControlType } from './helpers/make-api-call';

export class APIError extends Error {
  code: number;

  constructor({ Control }: { Control: ControlType}) {
    super(Control.Message || 'Unexpected response from API');
    this.code = Control.MessageCode;
  }
}

export class HTTPError extends Error {
  status: number;

  constructor(response: Response) {
    super(`${response.status} - ${response.statusText}`);
    this.status = response.status;
  }
}
