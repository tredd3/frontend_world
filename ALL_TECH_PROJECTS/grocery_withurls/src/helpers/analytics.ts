import { getMobileOperatingSystem } from './utilities';
import { getUser } from '../services/user';
import { getCart, PaymentMethodId, paymentLabelById } from '../services/cart';

import { CheckoutData } from '../services/checkout';
import { CartPivot } from '../types';
import { flatMap } from './typed-utils';

let purchaseJourney: string;

export const setPurchaseJourney = (val: string) => {
  purchaseJourney = val;
};

export const getPurchaseJourney = () => purchaseJourney;

declare global {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface Window {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    digitalData: Record<string, any>;
    _satellite: {
      track: (event: string) => void;
    };
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const track = async (event: string, data: Record<string, any>) => {
  if (typeof window === 'undefined') return;
  const { userId } = await getUser();
  const userOS = getMobileOperatingSystem();
  const platform = userOS === 2 ? 'Android App' : userOS === 3 ? 'iOS App' : 'Web';

  const digitalData = {
    link: {
      ...data.link
    },
    internal: {
      ...data.internal
    },
    page: {
      pageInfo: {
        appName: 'JioMart Web'
      },
      ...data.page
    },
    user: {
      platform,
      loginStatus: 'Logged in',
      martId: userId,
      jioCenter: '',
      ...data.user
    },
    product: [
      ...(data.product ? data.product : [])
    ],
    interPromotion: {
      ...data.interPromotion
    },
    ...(data.product ? data.filterSection : {}),
    ...(data.purchase ? data.purchase : {})
  };

  window.digitalData = {
    ...window.digitalData,
    ...digitalData
  };

  // eslint-disable-next-line no-underscore-dangle
  if (window._satellite) window._satellite.track(event);
};

type LinkPositions = 'Top' | 'Middle' | 'Bottom';

export const trackLink = (linkName: string, linkType: string, linkPosition: LinkPositions) => (
  track('linkTracking', { link: { linkName, linkType, linkPosition } })
);

export const trackPage = (pageName: string) => (
  track('pageLoad', { page: { pageName } })
);

type FilterSection = {
  filterSelection: boolean;
  sortSelection: boolean;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const trackProduct = (eventName: string, productData: Record<string, any>, filterSection?: FilterSection) => {
  const defaultFilterSection = {
    filterSelection: false,
    sortSelection: false
  };
  return track(eventName, {
    product: [productData],
    filterSection: (filterSection || defaultFilterSection)
  });
};

export const trackAPIShipments = (eventName: string, shipments: CheckoutData['Shipments'], purchase?: any) => {
  if (shipments && shipments.length) {
    track(eventName, {
      product: flatMap(shipments, shipment => shipment.Product)
        .map(product => ({
          productID: product.ProductSkuId,
          quantity: product.SelectedQuantity,
          price: product.SelectedQuantity * product.SP,
          productName: product.ProductName
        })),
      ...(purchase && { purchase })
    });
  }
};

type PurchaseArg = {
  paymentMethod: PaymentMethodId;
  paymentStatus?: 'Paid' | 'Unpaid';
  cart: CartPivot;
  failureReason?: string;
}

export const trackShipments = async (eventName: string, purchase?: PurchaseArg) => {
  const { shipments } = purchase?.cart || await getCart();
  if (!shipments || !shipments.length) return Promise.resolve();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let additionalInfo: Record<string, any> | null = null;

  if (purchase) {
    const { paymentMethod, cart, ...rest } = purchase;
    additionalInfo = {
      paymentMethod: paymentLabelById(paymentMethod),
      couponDiscount: cart.couponDiscountAmount,
      deliveryCharges: cart.deliveryCharge,
      totalOrderValue: cart.totalPayableAmount,
      transactionId: cart.cartId,
      ...rest
    };
  }

  return track(eventName, {
    product: flatMap(shipments, shipment => shipment.products)
      .map(product => ({
        productId: product.skuId,
        quantity: product.selectedQuantity,
        price: product.selectedQuantity * product.sp,
        productName: product.name
      })),
    ...(additionalInfo && additionalInfo)
  });
};
