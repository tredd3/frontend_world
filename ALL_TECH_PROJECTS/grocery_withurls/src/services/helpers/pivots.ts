import {
  APIShipment, APIShipmentProduct, APICart, APIProduct, ShipmentProductPivot,
  CartPivot, ProductPivot, couponDiscount, APIDashboardProduct,
  DashboardProductPivot, APIUser, UserPivot, AddressPivot, ShipmentPivot,
  APIStorePickup, APIShipmentProductCoupon, DeliveryModePivot, ConfigPivot,
  APIConfig, LastOrderDetails, Notification, NotificationPivot
} from '../../types';
import { genderCodeToString } from './generic';
import { APIAddressRequest } from '../address';

export const pivotShipmentProductCoupon = (coupon: APIShipmentProductCoupon) => ({
  id: coupon.CouponId,
  code: coupon.CouponCode,
  discountAmount: coupon.Discount,
  title: coupon.Title,
  description: coupon.Description
});

export const pivotShipmentProduct = (product: APIShipmentProduct): ShipmentProductPivot => ({
  mrp: product.MRP,
  sp: product.SP,
  otherImages: product.ProductOtherImage,
  name: product.ProductName,
  image: product.ProductImage,
  skuId: product.ProductSkuId,
  id: product.ProductId,
  isFc: product.IsFc,
  categoryId: product.CategoryId,
  selectedQuantity: product.SelectedQuantity,
  savingsPercent: product.PricePercentage,
  coupon: product.Coupons && pivotShipmentProductCoupon(product.Coupons)
});

export const pivotShipment = (shipment: APIShipment): ShipmentPivot => ({
  type: shipment.Type,
  deliveryDate: shipment.DeliveryDate,
  totalAmount: shipment.TotalAmount,
  discountAmount: shipment.TotalDiscount,
  couponDiscountAmount: shipment.TotalCouponDiscount,
  totalPayable: shipment.TotalPay,
  deliveryCharge: shipment.DeliveryCharge,
  deliveryDay: shipment.DeliveryDay,
  creationDate: shipment.CreationDate,
  itemCount: shipment.TotalItem,
  products: shipment.Product.map(pivotShipmentProduct)
});

export const pivotProduct = (product: APIProduct): ProductPivot => ({
  mrp: product.MRP,
  sp: product.SP,
  name: product.ProductName,
  images: (product.ProductOtherImage && product.ProductOtherImage.length) ? product.ProductOtherImage : [product.ProductImage],
  id: product.ProductId,
  skuId: product.ProductSkuId,
  isFc: product.IsFc,
  category: {
    name: product.CategoryName,
    id: product.CategoryId
  },
  alias: product.ProductAlias,
  description:
    product.SkuBulletPoint && product.SkuBulletPoint.length
      ? `${product.Description}. ${product.SkuBulletPoint.reduce(
        (acc, cv) => `${acc}. ${cv}`
      )}`
      : product.Description,
  disclaimer: product.Disclaimer,
  bestBefore: product.BestBeforeDate ? new Date(Number(product.BestBeforeDate)) : null,
  isVeg: product.FoodType ? Boolean(product.FoodType.toLocaleLowerCase() === 'veg') : null,
  additionalInformation: product.AdditionalInfo && product.AdditionalInfo !== 'null' ? product.AdditionalInfo : '',
  manufacturerAddress: product.ManufacturerAddress && product.ManufacturerAddress !== 'null' ? product.ManufacturerAddress : '',
  coupon: product.Coupons && Object.keys(product.Coupons).length > 0 ? {
    title: product.Coupons.Title,
    description: product.Coupons.Description,
    discountAmount: product.Coupons.DiscountType === couponDiscount.percentageDiscount
      ? product.Coupons.DiscountMax
      : product.Coupons.DiscountAbsValue
  } : null,
  brandName: product.BrandName,
  deliverySlot: product.DeliverySlot,
  ingredients: product.Ingredient,
  nutritionalFacts: product.NutritionalFacts
});

export const pivotCartResponse = (cartData: APICart): CartPivot => ({
  cartId: cartData.CartId,
  totalAmount: cartData.TotalAmount,
  discountAmount: cartData.TotalMRPDiscount,
  deliveryCharge: cartData.TotalDeliveryCharge,
  itemCount: cartData.TotalItem,
  couponDiscountAmount: cartData.TotalCouponDiscount,
  totalPayableAmount: cartData.TotalPay,
  shipments: cartData.Shipments.map(pivotShipment),
  savedForLater: cartData.SaveLater?.length ? cartData.SaveLater.map(pivotShipmentProduct) : null
});

export const pivotDashboardProduct = (product: APIDashboardProduct): DashboardProductPivot => ({
  brandName: product.BrandName,
  id: product.ProductId,
  images: product.ProductOtherImage?.length ? product.ProductOtherImage : [product.ProductImage],
  isFc: product.IsFc,
  mrp: product.MRP,
  name: product.ProductName,
  skuId: product.ProductSkuId,
  sp: product.SP,
  category: {
    name: product.CategoryName,
    id: product.CategoryId
  },
  coupon: product.CouponResult && Object.keys(product.CouponResult).length > 0 ? {
    title: product.CouponResult.Title,
    description: product.CouponResult.Description,
    discountAmount: product.CouponResult.DiscountType === couponDiscount.percentageDiscount
      ? product.CouponResult.DiscountMax
      : product.CouponResult.DiscountAbsValue
  } : null
});

export const pivotAPIUser = (user: APIUser): UserPivot => {
  const {
    mobileNumber, emailId, gender, dob, userId, ...otherUserProps
  } = user;
  const { latitude: lat, longitude: lng, ...remainingPreferences } = otherUserProps.preferences || ({} as UserPivot['preferences']);

  return {
    phoneNumber: mobileNumber,
    email: emailId,
    userId,
    gender: genderCodeToString(gender),
    ...(dob && { dob: new Date(dob) }),
    ...otherUserProps,
    preferences: { ...remainingPreferences, lat, lng },
    login: (userId !== undefined && userId !== null)
  };
};

export const pivotAPIAddress = ({ addressId, addedDate, ...address }: APIAddressRequest): AddressPivot => ({
  id: addressId,
  addedDate: new Date(addedDate.replace(/(A|P)M IST/, '')),
  ...address
});

export const pivotDeliveryMode = (date: APIStorePickup): DeliveryModePivot => ({
  id: date.DateId,
  label: date.Date,
  value: date.DateValue,
  isOpen: date.OpenStatus === 1,
  store: {
    isKdp: Boolean(date.KdpStore),
    name: date.Store.Name,
    storeId: Number(date.Store.StoreId)
  }
});

export const pivotConfig = (apiConfig: APIConfig): ConfigPivot => {
  let lastOrderDetails = null;
  if (apiConfig.LAST_ORDER_DETAILS && Object.keys(apiConfig.LAST_ORDER_DETAILS).length !== 0) {
    const lod = apiConfig.LAST_ORDER_DETAILS as LastOrderDetails;
    lastOrderDetails = {
      totalItems: Number(lod.TOTAL_ITEMS),
      orderId: Number(lod.ORDER_ID),
      storeName: lod.STORE_NAME,
      totalShipment: lod.TOTAL_SHIPMENT,
      deliveredDate: lod.DELIVERED_DATE,
      storeId: lod.STORE_ID,
      orderAmount: lod.ORDER_AMOUNT,
      deliveredTime: lod.DELIVERED_TIME
    };
  }

  return {
    pageSize: Number(apiConfig.DEFAULT_PAGE_SIZE),
    quantityErrorMessage: apiConfig.QTY_ERROR_MESSAGE,
    invalidOTPMessage: apiConfig.INVALID_OTP,
    maxCODLimit: Number(apiConfig.MAX_COD_LIMIT),
    deliveryChargeThreshold: Number(apiConfig.DELIVERY_CHARGE_THRESHOLD_AMOUNT),
    unreadNotificationCount: Number(apiConfig.UNREAD_NOTIFICATION_COUNT),
    deliveryCharges: Number(apiConfig.DELIVERY_CHARGES),
    codLimitErrorMessage: apiConfig.COD_LIMIT_ERROR_MESSAGE,
    welcomeProductSKU: Number(apiConfig.WELCOME_PRODUCT_SKU),
    genericErrorMessage: apiConfig.GENERIC_ERROR_MESSAGE,
    maxQuantityLimitPerItem: Number(apiConfig.MAX_QTY_LIMIT_PER_ITEM),
    lastOrderDetails
  };
};

export const pivotNotification = (notification: Notification): NotificationPivot => ({
  id: notification.notificationId,
  title: notification.title,
  body: notification.body,
  orderId: notification.data?.orderId ? notification.data.orderId : undefined,
  date: new Date(notification.createdTime),
  icon: notification.icon,
  status: notification.status.toLowerCase() as 'read' | 'unread'
});
