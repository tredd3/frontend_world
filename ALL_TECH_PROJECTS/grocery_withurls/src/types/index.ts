export * from './generic';
export * from './api';
export * from './pivot';
