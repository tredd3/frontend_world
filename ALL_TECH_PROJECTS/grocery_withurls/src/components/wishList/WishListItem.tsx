/** @jsx jsx */
import React from 'react';
import { jsx, css } from '@emotion/core';
import { ProductPivot } from '../../types';
import Price from '../uiControls/Price';
import Button from '../uiControls/button';
import { trackProduct } from '../../helpers/analytics';

const styles = {
  container: css`
    display: flex;
    padding: 12px 0;
  `,
  image: (url: string) => css`
    background-image: url(${url});
    background-repeat: no-repeat;
    background-position: center;
    background-size: contain;
    width: 20%;
    margin-right: 20px;
  `,
  detailsContainer: css`
    width: 80%;
  `,
  productName: css`
    width: 90%;
    font-size: 13px;
  `,
  detailsSection: css`
    display: flex;
    justify-content: space-between;
  `,
  priceWrapper: css`
    width: 65%;
    padding-top: 12px;
  `
};

type Props = {
  product: ProductPivot;
  onClickItem: (skuId: number) => void;
  onClickDelete: (skuId: number) => void;
  onClickAdd: (product: ProductPivot) => void;
};

const WishListItem: React.FC<Props> = ({
  product,
  onClickItem,
  onClickDelete,
  onClickAdd
}) => {
  const {
    category,
    images,
    skuId,
    sp,
    mrp,
    brandName,
    name
  } = product;

  const sendAnalytics = (eventName: string) => {
    trackProduct(eventName, {
      productID: skuId,
      quantity: 1,
      price: mrp,
      productCategory: category.name,
      purchaseJourney: 'Wishlist',
      brandName,
      productName: name
    });
  };

  return (
    <div
      css={styles.container}
      onClick={() => {
        onClickItem(skuId);
      }}
    >
      <div css={styles.image(images[0])} />
      <div css={styles.detailsContainer}>
        <div css={styles.productName}>{name}</div>

        <section css={styles.detailsSection}>
          <div css={styles.priceWrapper}>
            <Price sp={sp} mrp={mrp} />
          </div>
        </section>

        <div style={{ display: 'flex', justifyContent: 'space-between', marginRight: '10px' }}>
          <Button
            text="Delete"
            name="deleteFromCart"
            style={{
              padding: '6px', paddingLeft: '25px', paddingRight: '25px', marginTop: '6px'
            }}
            onClick={(e: React.MouseEvent) => {
              e.stopPropagation();
              sendAnalytics('removeFromWishlist');
              onClickDelete(skuId);
            }}
            type="solidGray"
          />
          <Button
            text="Add to cart"
            name="addToCart"
            style={{
              padding: '6px', paddingLeft: '18px', paddingRight: '18px', marginTop: '6px'
            }}
            onClick={(e: React.MouseEvent) => {
              e.stopPropagation();
              sendAnalytics('addToCart');
              onClickAdd(product);
            }}
            type="solidGray"
          />
        </div>
      </div>
    </div>
  );
};

export default WishListItem;
