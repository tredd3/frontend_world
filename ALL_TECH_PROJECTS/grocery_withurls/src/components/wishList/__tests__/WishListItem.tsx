import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { pivotProduct } from '../../../services/helpers/pivots';
import WishListItem from '../WishListItem';
import { APIProduct } from '../../../types';
import * as analytics from '../../../helpers/analytics';

const mockedItem = {
  MRP: 60,
  SP: 58.8,
  CategoryName: 'Fruit Juice',
  ProductOtherImage: [
    'https://api.reliancesmart.in//images/?name=images/hbf/h68/9006373404702.jpg',
    'https://api.reliancesmart.in//images/?name=images/h2c/hbb/9006338408478.jpg',
    'https://api.reliancesmart.in//images/?name=images/h84/h32/9006353350686.jpg',
    'https://api.reliancesmart.in//images/?name=images/h5c/hae/9006365081630.jpg',
    'https://api.reliancesmart.in//images/?name=images/he2/h0b/9006370127902.jpg'
  ],
  ProductName: 'Kato Grape Juice with Nata De Coco 330 ml',
  StoreProductId: '1000001710092643_fc',
  ProductImage: 'https://api.reliancesmart.in//images/?name=images/hbf/h68/9006373404702.jpg',
  StorePincode: '410210',
  FcId: 'VS33',
  ProductSkuId: 10092643,
  ProductSkuName: '1',
  BrandName: 'Kato',
  ProductId: 100081004,
  ProductLastUpdatedDate: '1577748676000',
  IsKA: 0,
  IsFc: 1,
  BrandId: 10004969,
  IsUKA: 0,
  isDCKA: 0,
  CategoryId: 10001089,
  Barcode: '490606980',
  CategoryIds: [
    1,
    10001761,
    10001770,
    10001089
  ],
  CategoryNames: [
    'Kirana / Provision store / Supermarket',
    'Health Drinks & Beverages',
    'Soft Drinks',
    'Fruit Juice'
  ],
  CGST: '0.0',
  IGST: '0.0',
  SGST: '0.0',
  CESS: '0.0',
  ADC: '0.0',
  IndustryId: '1',
  CompanyName: 'Taveephol Product Co., Ltd.',
  ProductAlias: 'Kato Grape Juice with Nata De Coco 330 ml 1 pcs',
  IsAvailable: 1,
  // eslint-disable-next-line max-len
  Description: 'Feel refreshed with the heavenly taste of Kato Kato Grape Juice with Nata De Coco. It tastes amazing, gives you an instant energy kick and quenches your thirst. It is also made from the finest quality of grapes and combines quality and affordability at the same time. So whether you are a child or an adult, this grape juice will satisfy your Grape cravings any time of the year. So go ahead, buy Kato Grape Juice with Nata De Coco online today!',
  ManufacturerAddress: 'null',
  // eslint-disable-next-line max-len
  Disclaimer: 'Despite our attempts to provide you with the most accurate information possible, the actual packaging, ingredients and colour of the product may sometimes vary. Please read the label, directions and warnings carefully before use.',
  SkuBulletPoint: [
    'Grape Flavour'
  ],
  BestBeforeDate: '1591079400000',
  SelectedQuantity: 0,
  FCProductId: '1000000000611413',
  UQOCID: '6',
  UQOCNAME: 'Pieces',
  UQCSHORTNAME: 'pcs'
};

describe('WishListItem component unit tests', () => {
  it('should render the component correctly', () => {
    const trackProductAnalyticsSpy = jest.spyOn(analytics, 'trackProduct').mockImplementation(jest.fn());
    const onClickItemSpy = jest.fn();
    const onClickAddSpy = jest.fn();
    const onClickDeleteSpy = jest.fn();

    const pivotedProduct = pivotProduct(mockedItem as unknown as APIProduct);

    const { asFragment, container, getByText } = render(<WishListItem
      product={pivotedProduct}
      onClickItem={onClickItemSpy}
      onClickAdd={onClickAddSpy}
      onClickDelete={onClickDeleteSpy}
    />);

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(container.firstChild as HTMLElement);
    expect(onClickItemSpy).toHaveBeenCalledTimes(1);
    expect(onClickItemSpy).toHaveBeenLastCalledWith(mockedItem.ProductSkuId);
    expect(trackProductAnalyticsSpy).not.toHaveBeenCalled();

    fireEvent.click(getByText('Add to cart'));
    expect(onClickAddSpy).toHaveBeenCalledTimes(1);
    expect(onClickAddSpy).toHaveBeenLastCalledWith(pivotedProduct);
    expect(trackProductAnalyticsSpy).toHaveBeenLastCalledWith('addToCart', {
      productID: mockedItem.ProductSkuId,
      quantity: 1,
      price: mockedItem.MRP,
      productCategory: mockedItem.CategoryName,
      purchaseJourney: 'Wishlist',
      brandName: mockedItem.BrandName,
      productName: mockedItem.ProductName
    });

    fireEvent.click(getByText('Delete'));
    expect(onClickDeleteSpy).toHaveBeenCalledTimes(1);
    expect(onClickDeleteSpy).toHaveBeenLastCalledWith(mockedItem.ProductSkuId);
    expect(trackProductAnalyticsSpy).toHaveBeenLastCalledWith('removeFromWishlist', {
      productID: mockedItem.ProductSkuId,
      quantity: 1,
      price: mockedItem.MRP,
      productCategory: mockedItem.CategoryName,
      purchaseJourney: 'Wishlist',
      brandName: mockedItem.BrandName,
      productName: mockedItem.ProductName
    });
  });
});
