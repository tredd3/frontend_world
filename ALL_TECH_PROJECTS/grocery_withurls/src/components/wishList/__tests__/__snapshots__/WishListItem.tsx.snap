// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishListItem component unit tests should render the component correctly 1`] = `
<DocumentFragment>
  .emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 12px 0;
}

.emotion-0 {
  background-image: url(https://api.reliancesmart.in//images/?name=images/hbf/h68/9006373404702.jpg);
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 20%;
  margin-right: 20px;
}

.emotion-8 {
  width: 80%;
}

.emotion-1 {
  width: 90%;
  font-size: 13px;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-6 {
  width: 65%;
  padding-top: 12px;
}

.emotion-5 {
  font-size: 10px;
  margin-bottom: 6px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  color: #A82205;
  margin-right: 5px;
  font-size: 16px;
  font-family: inherit;
  font-weight: 530;
}

.emotion-3 {
  color: #999;
  margin-right: 5px;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  font-size: 12px;
  font-family: inherit;
}

.emotion-4 {
  color: #616267;
  margin-right: 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 10px;
  font-family: inherit;
}

<div
    class="emotion-9"
  >
    <div
      class="emotion-0"
    />
    <div
      class="emotion-8"
    >
      <div
        class="emotion-1"
      >
        Kato Grape Juice with Nata De Coco 330 ml
      </div>
      <section
        class="emotion-7"
      >
        <div
          class="emotion-6"
        >
          <div
            class="emotion-5"
          >
            <span
              class="emotion-2"
            >
              ₹58.80
            </span>
            <br />
            <span
              class="emotion-3"
            >
              ₹60
            </span>
            <span
              class="emotion-4"
            >
              Save ₹1.20 (2%)
            </span>
          </div>
        </div>
      </section>
      <div
        style="display: flex; justify-content: space-between; margin-right: 10px;"
      >
        <div
          class="Component-button-3 Component-solidGray-4 solidGray Component-enable-8"
        >
          <button
            name="deleteFromCart"
            style="padding: 6px 25px 6px 25px; margin-top: 6px;"
            type="button"
          >
            Delete
          </button>
        </div>
        <div
          class="Component-button-3 Component-solidGray-4 solidGray Component-enable-8"
        >
          <button
            name="addToCart"
            style="padding: 6px 18px 6px 18px; margin-top: 6px;"
            type="button"
          >
            Add to cart
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
